### ISSUE TITLE
Defining extension trait for `f32::sin()` in `#![no_std]` program produces incorrect warning

### ISSUE THREAD

**Issue by khyperia** [rust-lang/rust#80837]
Originally found here: https://twitter.com/khyperia/status/1346902124806942730

`a/src/lib.rs`:
```rust
pub trait Ext {
    fn sin(self) -> Self;
}
impl Ext for f32 {
    fn sin(self) -> Self { 0.0 }
}
```
`b/src/lib.rs`:
```rust
#![no_std]
use a::Ext;
pub fn func() -> f32 {
    1.0f32.sin()
}
```

**Note the lack of `#![no_std]` in `a/src/lib.rs`**

<details>
  <summary>both `a` and `b` are typical crates, created with this (click to expand)</summary>

```
rusttest> mkdir a
rusttest> cd a
rusttest/a> cargo init --vcs none --lib
     Created library package
rusttest/a> mkdir ../b
rusttest/a> cd ../b
rusttest/b> cargo init --vcs none --lib
     Created library package
rusttest/b> nvim Cargo.toml
...
[dependencies]
a = { path = "../a" }
```
</details>

I expected to see this happen: Because crate `b` is `#![no_std]`, `f32::sin()` does not exist (it only exists in `std`). So, it needs `a::Ext::sin()` to be imported. I expect the code to compile without warnings.

Instead, this happened: An **unused import warning** is reported on `use a::Ext;`. This is incorrect, deleting the `use` cases a compilation failure with `no method named sin() found` (as it should).

### Meta

`rustc --version --verbose`:
```
rustc 1.48.0 (7eac88abb 2020-11-16)
binary: rustc
commit-hash: 7eac88abb2e57e752f3302f02be5f3ce3d7adfb4
commit-date: 2020-11-16
host: x86_64-pc-windows-msvc
release: 1.48.0
LLVM version: 11.0
```
`rustc +nightly --version --verbose`:
```
rustc 1.50.0-nightly (1c389ffef 2020-11-24)
binary: rustc
commit-hash: 1c389ffeff814726dec325f0f2b0c99107df2673
commit-date: 2020-11-24
host: x86_64-pc-windows-msvc
release: 1.50.0-nightly
```
(reproduces on both nightly and stable, although I haven't updated in a while...)

---

This may be related to some issues @bitshifter was seeing in [glam](https://github.com/bitshifter/glam-rs), so pinging them if they'd like to follow this, as well as embark on rust-gpu (which I'm on).

Context here is that we want methods like `sin()` and all the other "advanced" math methods on `f32` in `no_std` programs, which are not available, so we're resorting to [rust libm](https://github.com/rust-lang/libm) and extension traits (and `num-traits` serves that need as a glorified extension trait that calls through to `libm`). This warning pops up all the time and it's really frustrating to try and figure out what's happening and why.

**Comment by khyperia** [comment]
@m-ou-se pointed out that this isn't related to the extension trait, but rather `std` getting pulled in:

`a/src/lib.rs`:
```rust
pub fn thing(){}
```
`b/src/lib.rs`:
```rust
#![no_std]
use a::thing;
pub fn func() -> f32 {
    1.0f32.sin()
}
```

causes `unused import: a::thing` but still succeeds compilation, deleting `use a::thing;` fails with `sin() not found`. So, still an incorrect warning I guess?

**Comment by parasyte** [comment]
A better diagnostic would recommend something like adding `#[cfg(not(test))]` to the import:

```rust
#[cfg(not(test))]
use num_traits::float::FloatCore as _;
```

This is at least the case I have, where I am running tests (on the host) for my `no_std` lib, and the automatic `std::prelude` creates this conflict with the `test` profile. Conditionally disabling the import for the `test` configuration is easy does the trick.
